exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,421 @@
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSetBinding object after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSetBinding); err != nil {
+			r.Log.Error(err, "failed to patch config")
+		}
+	}()
+
+	// Initializes the maps in ClusterResourceSetBinding object if they are uninitialized.
+	initClusterResourceSetBinding(clusterResourceSetBinding, clusterResourceSet)
+
+	var clusterResourceSetInMap addonsv1.ResourcesSetBinding
+	var ok bool
+	errList := make([]error, 0)
+	clusterResourceSetInMap = clusterResourceSetBinding.Spec.Bindings[clusterResourceSet.Name]
+
+	// Iterate all resources and apply them to the cluster and update the resource status in the ClusterResourceSetBinding object.
+	for _, resource := range clusterResourceSet.Spec.Resources {
+		// If resource is already applied successfully and clusterResourceSet mode is "ApplyOnce", continue. (No need to check hash changes here)
+		resourceInMap := addonsv1.ResourceBinding{}
+		if resourceInMap, ok = clusterResourceSetInMap.Resources[generateResourceKindName(resource)]; ok {
+			if clusterResourceSet.Spec.Strategy == string(addonsv1.ClusterResourceSetStrategyApplyOnce) && resourceInMap.Applied {
+				continue
+			}
+		}
+
+		// This is set here to cover unsuccessful apply scenarios (if the end of the loop is not reached)
+		resourceInMap = generateResourceStatus(false, "")
+		clusterResourceSetInMap.Resources[generateResourceKindName(resource)] = resourceInMap
+
+		// Only allow using resources (Secrets/Configmaps) in the same namespace with the cluster.
+		resourceName := types.NamespacedName{Name: resource.Name, Namespace: cluster.Namespace}
+		dataList := make([][]byte, 0)
+
+		// Retrieve data in the resource as an array because there can be many key-value pairs in resource and all will be applied to the cluster.
+		switch resource.Kind {
+		case string(addonsv1.ConfigMapClusterResourceSetResourceKind):
+			// Set status in ClusterResourceSetBinding in case of early continue due to a failure. Do this for only supported kinds.
+			clusterResourceSetInMap.Resources[generateResourceKindName(resource)] = addonsv1.ResourceBinding{
+				Hash:            "",
+				Applied:         false,
+				LastAppliedTime: &metav1.Time{Time: time.Now().UTC()},
+			}
+
+			resourceConfigMap, err := getConfigMap(context.Background(), r.Client, resourceName)
+			if err != nil {
+				logger.Error(err, "Failed to find ConfigMap resource", "ConfigMap name", resourceName)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.MissingResourceReason, clusterv1.ConditionSeverityWarning, err.Error())
+				errList = append(errList, err)
+				continue
+			}
+
+			// Convert the object to unstructured to compare against our before copy.
+			raw, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resourceConfigMap.DeepCopyObject())
```

Comment:
This is to avoid having 2 different methods for Secret resource and ConfigMap resource.

Reply:
Add a comment to that effect :)

Reply:
Just noticed that even though we're converting it to unstructured, we're duplicating the same code here and in the other switch, we should avoid doing so if the goal was to avoid repetition